### PR TITLE
Schemaless Browser Tool

### PR DESCRIPTION
**Pull request by pankti-m**
Schemaless Browser Tool using Dynamic Messages.

Description:

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/collections/DynamicCorfuKey.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        this.payload = new CorfuDynamicMessage(payload);
+    }
+
+    public DynamicMessage getPayload() {
+        return payload.getPayload();
+    }
+
+    @Override
+    public int hashCode() {
+        return payload.hashCode();
```
Comment (pankti-m) [inline runtime/src/main/java/org/corfudb/runtime/collections/DynamicCorfuKey.java]: shouldnt the type url be factored in to calculate the hashcode?

File: runtime/src/main/java/org/corfudb/runtime/collections/DynamicCorfuKey.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null) {
+            return false;
+        }
+        DynamicCorfuKey dynamicCorfuKey = (DynamicCorfuKey) obj;
```
Comment (pankti-m) [inline runtime/src/main/java/org/corfudb/runtime/collections/DynamicCorfuKey.java]: same here.  Shouldnt the type url be factored in to determine equality?

File: runtime/src/main/java/org/corfudb/runtime/collections/DynamicCorfuKey.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        return payload.equals(dynamicCorfuKey.payload);
```
Comment (pankti-m) [inline runtime/src/main/java/org/corfudb/runtime/collections/DynamicCorfuKey.java]: same here. Shouldnt the type url be factored in to determine equality?

File: runtime/src/main/java/org/corfudb/util/serializer/DynamicProtobufSerializer.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        }
+    }
+
+    @Override
+    public byte getType() {
+        return type;
+    }
+
+    FileDescriptor getDescriptors(String name) throws DescriptorValidationException {
+
```
Comment (pankti-m) [inline runtime/src/main/java/org/corfudb/util/serializer/DynamicProtobufSerializer.java]: nit - extra newline

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import org.corfudb.runtime.collections.DynamicCorfuKey;
+import org.corfudb.runtime.collections.DynamicCorfuRecord;
+import org.corfudb.util.serializer.DynamicProtobufSerializer;
+import org.corfudb.util.serializer.ISerializer;
+import org.corfudb.util.serializer.Serializers;
+
+@Slf4j
+public class CorfuBrowser {
+    private CorfuRuntime runtime;
+    String singleNodeEndpoint;
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: private final

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    public CorfuBrowser(String host, int port) {
+        singleNodeEndpoint = String.format("%s:%d", host, port);
+        runtime = createRuntime(singleNodeEndpoint);
+    }
+
+    private CorfuRuntime createRuntime(String endpoint) {
+        runtime = new CorfuRuntime(endpoint).setCacheDisabled(true).connect();
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: This is a deprecated constructor. Please use CorfuRuntimeParameters.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        return runtime;
+    }
+
+    public CorfuTable getTable(String namespace, String tableName) {
+        final int code = 25;
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: This number will get difficult to keep track of. We should keep this at some common location.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        ISerializer dynamicProtobufSerializer =
+            new DynamicProtobufSerializer((byte) code, runtime);
+        Serializers.registerSerializer(dynamicProtobufSerializer);
+
+        CorfuTable<DynamicCorfuKey, DynamicCorfuRecord> corfuTable =
+            runtime.getObjectsView().build()
+            .setTypeToken(new TypeToken<CorfuTable<DynamicCorfuKey, DynamicCorfuRecord>>() {
+            })
+            .setStreamName(namespace + "$" + tableName)
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: Use `TableRegistry.getFullyQualifiedTableName`

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+    public void printTable(CorfuTable table) {
+        for (Object obj : table.values()) {
+            DynamicCorfuRecord record = (DynamicCorfuRecord)obj;
+
+            log.info("Payload: {}", record.getPayload());
+            log.info("Metadata: {}", record.getMetadata());
+            log.info("===================");
+
+            System.out.println("Payload: " + record.getPayload());
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: Remove System outs.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+import org.corfudb.util.GitRepositoryState;
+import org.docopt.Docopt;
+
+public class CorfuBrowserMain {
+    private static final String USAGE = "Usage: corfu-browser --host=<host> " +
+        "--port=<port> --namespace=<namespace> --tablename=<tablename>\n"
+        + "Options:\n"
+        + "--host=<host>   Hostname\n"
+        + "--port=<port>   Port\n";
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: Add options for namespace and table name as well

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java (hunk 5 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: Done

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java (hunk 6 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: done.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java (hunk 7 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: Yes.  One way is to keep a configuration file with all constant parameters used by corfu.  Right now I have created a constant in this file itself.  I can work on this cleanup in parallel.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java (hunk 8 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: done

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java (hunk 10 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: done

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+import org.corfudb.runtime.collections.CorfuTable;
+import org.corfudb.runtime.collections.CorfuDynamicKey;
+import org.corfudb.runtime.collections.CorfuDynamicRecord;
+import org.corfudb.runtime.view.TableRegistry;
+import org.corfudb.util.serializer.DynamicProtobufSerializer;
+import org.corfudb.util.serializer.ISerializer;
+import org.corfudb.util.serializer.Serializers;
+
+@Slf4j
+public class CorfuBrowser {
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: Add Class level comments.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        connectRuntime(singleNodeEndpoint);
+    }
+
+    private void connectRuntime(String endpoint) {
+        runtime.parseConfigurationString(endpoint);
+        runtime.connect();
+        log.info("Successfully connected to {}", endpoint);
+    }
+
+    public CorfuTable getTable(String namespace, String tableName) {
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: Add comments on all methods.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+            .setTypeToken(new TypeToken<CorfuTable<CorfuDynamicKey, CorfuDynamicRecord>>() {
+            })
+            .setStreamName(
+                TableRegistry.getFullyQualifiedTableName(namespace, tableName))
+            .setSerializer(dynamicProtobufSerializer)
+            .open();
+        return corfuTable;
+    }
+
+    @SuppressWarnings("checkstyle:printLine")
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: Since this is only for testing. Remember to remove this before checking in.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.util.Map;
+
+import org.corfudb.util.GitRepositoryState;
+import org.docopt.Docopt;
+
+public class CorfuBrowserMain {
+    private static final String USAGE = "Usage: corfu-browser --host=<host> " +
+        "--port=<port> --namespace=<namespace> --tablename=<tablename> " +
+        "--keystore=<keystore_file> --ks_password=<keystore_password> " +
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: We can make these fields optional.
This will be helpful when we want to browse data on our local setup which might not have the tls requirements.
For optional parameters you can take a look at CorfuServer Docopt. (You need to enclose optional parameters in '[, ]')

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+        + "--port=<port>   Port\n"
+        + "--namespace=<namespace>   Namespace\n"
+        + "--tablename=<tablename>   Table Name\n"
+        + "--keystore=<keystore_file> KeyStore File\n"
+        + "--ks_password=<keystore_password> KeyStore Password\n"
+        + "--truststore=<truststore_file> TrustStore File\n"
+        + "--truststore_password=<truststore_password> Truststore Password\n"
+        + "--tlsEnabled=<tls_enabled>";
+
+    @SuppressWarnings("checkstyle:printLine")
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: Remove this after finishing testing.

File: test/src/test/java/org/corfudb/integration/CorfuBrowserIT.java [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+            "corfuSingleNodePort"));
+        singleNodeEndpoint = String.format(
+            "%s:%d",
+            corfuSingleNodeHost,
+            corfuStringNodePort
+        );
+    }
+
+    @Test
+    public void browserTest() throws
```
Comment (zalokhan) [inline test/src/test/java/org/corfudb/integration/CorfuBrowserIT.java]: Good tests.
Please add high-level comments describing the test.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java (hunk 11 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: done

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java (hunk 12 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: done

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java (hunk 13 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: done

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java (hunk 14 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: done

File: test/src/test/java/org/corfudb/integration/CorfuBrowserIT.java (hunk 16 above)
Comment (pankti-m) [reply test/src/test/java/org/corfudb/integration/CorfuBrowserIT.java]: done

**Comment by Maithem** [comment]
Can you rebase? there's a bunch of dependencies that were already merged.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+import org.corfudb.runtime.collections.CorfuTable;
+import org.corfudb.runtime.collections.CorfuDynamicKey;
+import org.corfudb.runtime.collections.CorfuDynamicRecord;
+import org.corfudb.runtime.view.TableRegistry;
+import org.corfudb.util.serializer.DynamicProtobufSerializer;
+import org.corfudb.util.serializer.ISerializer;
+import org.corfudb.util.serializer.Serializers;
+
+/**
+ * This is the UFO Browser Tool which prints data in a given namespace and table.
```
Comment (Maithem) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: Just to keep the terminology consistent. Maybe use CorfuTable instead ?

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+ *
+ * - Created by pmajmudar on 10/16/2019.
+ */
+@Slf4j
+public class CorfuBrowser {
```
Comment (Maithem) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: CorfuTableBrowser

**Comment by pankti-m** [comment]
Rebased.

~ Pankti

File: corfudb-tools/README [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+
+--------
+Browser
+--------
+The browser supports read operation which reads and outputs all entries in a
+given table and namespace.
+We plan to add a user script to invoke the browser but as of now, it can be used
+with the following command-
+
+java -cp <corfu-tools jar>
```
Comment (Maithem) [inline corfudb-tools/README]: I think the example should be invoked with actual arguments:
`java -cp corfudb-tools-0.3.0-SNAPSHOT-shaded.jar org.corfudb.browser.CorfuBrowserMain`

Since docopt generates the usage with --help, I don't think we also need to document them here

File: corfudb-tools/src/main/resources/logback.xml [hunk 20]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Maithem) [inline corfudb-tools/src/main/resources/logback.xml]: Why do you need both prod and this?

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+     * Creates a CorfuBrowser which connects a runtime to the server.
+     * @param host Corfu Server Host
+     * @param port Corfu Server Port
+     * @param keyStore Location of the KeyStore File
+     * @param ksPasswordFile Location of the KeyStore Password File
+     * @param trustStore Location of the Truststore File
+     * @param tsPasswordFile Location of the Truststore Password File
+     * @param tlsEnabled Tls Enabled Status
+     */
+    public CorfuBrowser(String host, int port, String keyStore,
```
Comment (Maithem) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: An endpoint is actually a cluster, not a single address. I suggest you just pass a comma separated connection string. For example, ip1:port1,ip2:port2

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    /**
+     * Setup logback logger
+     * - pick the correct logging level before outputting error messages
+     * - add serverEndpoint information
+     *
+     * @throws JoranException logback exception
+     */
+    private static void configureLogger() {
```
Comment (Maithem) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: Why is this needed, aren't you setting this from the *xml files?

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+import org.docopt.Docopt;
+
+/**
+ * Main class for the UFO Browser Tool.
+ * Command line options are documented in the USAGE variable.
+ *
+ * - Created by pmajmudar on 10/16/2019
+ */
+public class CorfuBrowserMain {
+    private static final String USAGE = "Usage: corfu-browser --host=<host> " +
```
Comment (Maithem) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: Btw, if you wan't an alternative to docopt, you can consider jcommander. I feel like its more cleaner way to specify command args.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+        String keystore = opts.get("--keystore") == null ? "" :
+            opts.get("--keystore").toString();
+        String ks_password = opts.get("--ks_password") == null? "" :
+            opts.get("--ks_password").toString();
+        String truststore = opts.get("--truststore") == null ? "" :
+            opts.get("--truststore").toString();
+        String truststore_password =
+            opts.get("--truststore_password") == null? "" :
+                opts.get("--truststore_password").toString();
+        boolean tlsEnabled = opts.get("--tlsEnabled") == null ? false :
```
Comment (Maithem) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: This might be superfluous, since you can actually pass nulls to the constructor.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+            Boolean.parseBoolean(opts.get("--tlsEnabled").toString());
+
+        CorfuBrowser browser = new CorfuBrowser(host, port, keystore,
+            ks_password, truststore, truststore_password, tlsEnabled);
+        CorfuTable table = browser.getTable(opts.get("--namespace").toString(),
+            opts.get("--tablename").toString());
+        browser.printTable(table);
```
Comment (Maithem) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: This seems to be crude. Why not have a REPL and support the following options:
1. Get key
2. print all keys
3. print the whole table

Comment (Maithem) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: If the table is too large and we're interested in only a couple of keys, its too much work to first dump the table in a separate file and then grep. It makes sense to support this operation directly.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    /**
+     * Prints the payload and metadata in the given table
+     * @param table
+     */
+    public void printTable(CorfuTable table) {
+        for (Object obj : table.values()) {
+            CorfuDynamicRecord record = (CorfuDynamicRecord)obj;
+            log.info("Payload: {}", record.getPayload());
```
Comment (Maithem) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: It will be nice to print this in table format using  width and precision specifiers
==========
Name: table name  Namespace: namespace
==========
Key   Value
k1      v1
k2      v2

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java (hunk 17 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: done.  replaced other references of UFO with CorfuTable.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java (hunk 18 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: done

File: corfudb-tools/README [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+
+The tool is schemaless, i.e., it does not need to be packaged with protobuf
+definitions or proto files to serialize/deserialize the data.
+
+--------
+Browser
+--------
+The browser supports read operation which reads and outputs all entries in a
+given table and namespace.
+We plan to add a user script to invoke the browser but as of now, it can be used
```
[the remark this answers is not in the document]
Comment (pankti-m) [reply corfudb-tools/README]: done

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java (hunk 23 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: We decided to use docopt because it is being used elsewhere in the repo.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java (hunk 24 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: I was not sure of the behavior of building CorfuRuntime parameters with null arguments.

If it does not support nulls, will have to add the same checks later.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java (hunk 25 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: for now we decided to start with a simple operation of just reading all the contents.  We plan to add more filtering capabilities later.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java (hunk 24 above)
Comment (WenbinZhu) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: Yea passing empty string to runtime parameter looks incorrect. You can pass null to it, and let it fail with NPE in case some parameter is required but not provided if tls is enabled. And if tls is not enabled, these parameters should be ignored.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+            opts.get("--truststore").toString();
+        String truststore_password =
+            opts.get("--truststore_password") == null? "" :
+                opts.get("--truststore_password").toString();
+        boolean tlsEnabled = opts.get("--tlsEnabled") == null ? false :
+            Boolean.parseBoolean(opts.get("--tlsEnabled").toString());
+
+        CorfuBrowser browser = new CorfuBrowser(host, port, keystore,
+            ks_password, truststore, truststore_password, tlsEnabled);
+        CorfuTable table = browser.getTable(opts.get("--namespace").toString(),
```
Comment (WenbinZhu) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: Does this mean a single invocation of this tool can only dump one table? If I want to dump multiple tables, I need to re-invoke the tool, create a new runtime and connect?

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+        return corfuTable;
+    }
+
+    /**
+     * Prints the payload and metadata in the given table
+     * @param table
+     */
+    public void printTable(CorfuTable table) {
+        for (Object obj : table.values()) {
+            CorfuDynamicRecord record = (CorfuDynamicRecord)obj;
```
Comment (WenbinZhu) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: Why we are only printing the record and metadata, not the keys?

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java (hunk 21 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: I based this on the current implementation of the browser tool in the NSX repo.  Will address this in a subsequent patch.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java (hunk 24 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: done

File: corfudb-tools/README [hunk 30]
```diff
@@ -0,0 +1,33 @@
+UFO Browser Editor Tool
```
[the remark this answers is not in the document]
Comment (pankti-m) [reply corfudb-tools/README]: I am changing it to CorfuStoreBrowser

File: corfudb-tools/README (hunk 19 above)
Comment (pankti-m) [reply corfudb-tools/README]: done

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java (hunk 28 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowserMain.java]: Yes a single invocation only prints contents of a single table.  Every invocation creates a new runtime and connects using it.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java (hunk 22 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: fixed

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 31]
```diff
@@ -0,0 +1,110 @@
...
+        runtime.connect();
+        log.info("Successfully connected to {}", endpoint);
+    }
+
+    /**
+     * Setup logback logger
+     * - pick the correct logging level before outputting error messages
+     * - add serverEndpoint information
+     *
+     * @throws JoranException logback exception
```
[the remark this answers is not in the document]
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: done

File: corfudb-tools/src/main/resources/logback.xml (hunk 20 above)
Comment (pankti-m) [reply corfudb-tools/src/main/resources/logback.xml]: changed to only use prod

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java (hunk 26 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: I tried this using string.format but to align the keys/values/metadata correctly in their respective columns, there is no certain width which is suitable for all types of data.  Without alignment, it is not readable.

So I am printing it in the form-
Key:
Value:
Meta:
<Row Separator>

Let me know if you want to see a sample output.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java [hunk 32]
```diff
@@ -0,0 +1,110 @@
...
+            .setSerializer(dynamicProtobufSerializer)
+            .open();
+        return corfuTable;
+    }
+
+    /**
+     * Prints the payload and metadata in the given table
+     * @param table
+     */
+    public void printTable(CorfuTable table) {
```
[the remark this answers is not in the document]
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: Added.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java (hunk 29 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuBrowser.java]: Added keys as well

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowser.java [hunk 33]
```diff
@@ -0,0 +1,80 @@
...
+        CorfuTable<CorfuDynamicKey, CorfuDynamicRecord> table) {
+        StringBuilder builder;
+        for (Map.Entry<CorfuDynamicKey, CorfuDynamicRecord> entry :
+            table.entrySet()) {
+            builder = new StringBuilder("\nKey:\n" + entry.getKey().getKey())
+                .append("\nPayload:\n" + entry.getValue().getPayload())
+                .append("\nMetadata:\n" + entry.getValue().getMetadata());
+                //.append("\n====================\n");
+            log.info(builder.toString());
+            /*log.info("\nKey: {}", entry.getKey().getKey());
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowser.java]: You can remove these comments.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowser.java [hunk 34]
```diff
@@ -0,0 +1,80 @@
...
+     */
+    public void printTable(
+        CorfuTable<CorfuDynamicKey, CorfuDynamicRecord> table) {
+        StringBuilder builder;
+        for (Map.Entry<CorfuDynamicKey, CorfuDynamicRecord> entry :
+            table.entrySet()) {
+            builder = new StringBuilder("\nKey:\n" + entry.getKey().getKey())
+                .append("\nPayload:\n" + entry.getValue().getPayload())
+                .append("\nMetadata:\n" + entry.getValue().getMetadata());
+                //.append("\n====================\n");
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowser.java]: I think you want to uncomment this right?

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowser.java [hunk 35]
```diff
@@ -0,0 +1,80 @@
...
+import org.corfudb.util.serializer.ISerializer;
+import org.corfudb.util.serializer.Serializers;
+
+/**
+ * This is the CorfuTable Browser Tool which prints data in a given namespace and table.
+ *
+ * - Created by pmajmudar on 10/16/2019.
+ */
+@Slf4j
+public class CorfuStoreBrowser {
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowser.java]: The comment says CorfuTableBrowser and class name is CorfuStoreBrowser.
Fix either of them.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserMain.java [hunk 36]
```diff
@@ -0,0 +1,81 @@
...
+
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.collections.CorfuDynamicKey;
+import org.corfudb.runtime.collections.CorfuDynamicRecord;
+import org.corfudb.runtime.collections.CorfuTable;
+import org.corfudb.util.GitRepositoryState;
+import org.docopt.Docopt;
+
+/**
+ * Main class for the CorfuTable Browser Tool.
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserMain.java]: Fix this to CorfuStore

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserMain.java [hunk 37]
```diff
@@ -0,0 +1,81 @@
...
+        + "--namespace=<namespace>   Namespace\n"
+        + "--tablename=<tablename>   Table Name\n"
+        + "--keystore=<keystore_file> KeyStore File\n"
+        + "--ks_password=<keystore_password> KeyStore Password\n"
+        + "--truststore=<truststore_file> TrustStore File\n"
+        + "--truststore_password=<truststore_password> Truststore Password\n"
+        + "--tlsEnabled=<tls_enabled>";
+
+    public static void main(String[] args) {
+        CorfuRuntime runtime;
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserMain.java]: Should we have a wrapping try-catch block?
What happens if there is an exception? Does it log or fail silently?

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserMain.java [hunk 38]
```diff
@@ -0,0 +1,81 @@
...
+        boolean tlsEnabled = Boolean.parseBoolean(opts.get("--tlsEnabled").toString());
+
+        if (tlsEnabled) {
+            String keystore = opts.get("--keystore").toString();
+            String ks_password = opts.get("--ks_password").toString();
+            String truststore = opts.get("--truststore").toString();
+            String truststore_password =
+                opts.get("--truststore_password").toString();
+            runtime = CorfuRuntime.fromParameters(
+                CorfuRuntime.CorfuRuntimeParameters.builder().useFastLoader(false)
```
Comment (zalokhan) [inline corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserMain.java]: Start the Parameters builder outside the if loop and build() it after the else to initialize the runtime.
You can then set duplicate options like fast loader and cache disabled only once above.

File: test/src/test/java/org/corfudb/integration/CorfuStoreBrowserIT.java [hunk 39]
```diff
@@ -0,0 +1,190 @@
...
+            .setLsb(metadataUuid)
+            .build();
+        TxBuilder tx = store.tx(namespace);
+        tx.create(tableName, uuidKey, uuidVal, metadata)
+            .update(tableName, uuidKey, uuidVal, metadata)
+            .commit();
+        runtime.shutdown();
+
+        // Invoke the browser and go through each item
+        runtime.connect();
```
Comment (zalokhan) [inline test/src/test/java/org/corfudb/integration/CorfuStoreBrowserIT.java]: Connecting an already shutdown runtime is not a good idea. Probably we should add a check to avoid that (Not in this PR).
Just create a new runtime.

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowser.java (hunk 34 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowser.java]: yes

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowser.java (hunk 35 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowser.java]: changed comment to 'CorfuStoreBrowser'

File: corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserMain.java (hunk 37 above)
Comment (pankti-m) [reply corfudb-tools/src/main/java/org/corfudb/browser/CorfuStoreBrowserMain.java]: runtime.connect() can throw an error and it gets logged.  Otherwise I did not find anything throwing an exception.  Maybe we can wrap this in a generic exception?

File: samples/src/main/java/org/corfudb/samples/CorfuStoreSample.java [hunk 40]
```diff
@@ -35,9 +35,16 @@
...
-                .build())
-                .parseConfigurationString(configurationString)
-                .connect();
+            .useFastLoader(false)
+            .cacheDisabled(true)
+            .tlsEnabled(true)
+            .keyStore("/config/cluster-manager/corfu/private/keystore.jks")
+            .ksPasswordFile("/config/cluster-manager/corfu/private/keystore.password")
+            .trustStore("/config/cluster-manager/corfu/public/truststore.jks")
+            .tsPasswordFile("/config/cluster-manager/corfu/public/truststore.password")
```
Comment (WenbinZhu) [inline samples/src/main/java/org/corfudb/samples/CorfuStoreSample.java]: Why the files are using these paths instead of a local file in the resource folder?

Comment (pankti-m) [reply samples/src/main/java/org/corfudb/samples/CorfuStoreSample.java]: Reverting to not using these parameters.

Doing so will not allow sample app to use tls but that is independent of browser changes.  So can be addressed in a different PR.
